tpm2/structures.go

```diff
@@ -137,6 +138,49 @@ func (p Public) Name() (*Name, error) {
 	}, nil
 }
 
+// MatchesTemplate checks if the Public area has the same algorithms and
+// parameters as the provided template. Note that this does not necessarily
+// mean that the key was created from this template, as the Unique field is
+// both provided in the template and overriden in the key creation process.
+func (p Public) MatchesTemplate(template Public) bool {
+	if p.Type != template.Type ||
+		p.NameAlg != template.NameAlg ||
+		p.Attributes != template.Attributes ||
+		!bytes.Equal(p.AuthPolicy, template.AuthPolicy) {
+		return false
+	}
+	switch p.Type {
+	case AlgRSA:
```

Comment:
How about moving type-specific logic into methods on those types?
e.g. code in this `case` would be `return p.RSAParameters.matchesTemplate(template.RSAParameters)`